### ISSUE TITLE
Class/Structure handling

### ISSUE THREAD

**Issue by faltur** [x64dbg/x64dbg#1108]
A suggestion to make our lives easier. 

When the user right-clicks a line on the disassembler window, the dbg shows the context menu option "Create class/structure for", and he will be able to select a register or a memory address. 
Example:

00A9F11F	        push ebp
00A9F120	mov ebp,esp
00A9F122	push ecx
00A9F123	mov dword ptr ss:[ebp-4],ecx     ; <<< the line selected by user
00A9F126	mov eax,dword ptr ss:[ebp-4]
00A9F129	xor ecx,ecx
00A9F12B	cmp dword ptr ds:[eax+1C],0
00A9F12F	        sete cl
00A9F132	mov al,cl
00A9F134	leave
00A9F135	ret

When the user selects the line 00A9F123 and right clicks it, he will be able to create a class/structure for ECX or EBP-4. A new dialog box is shown.

The "Create class/struct" dialog box contains the following input fields:
- an edit box to enter a new label for the selected address, and
- a multi-lined edit box (plainText) for the class/structure itself. The user can type a complete class or structure inside this field in C++. The name of the class/struct will be obtained from their definition.

Example of input class:

class MyClass
{
public:
	virtual void start();
	virtual void stop();
	virtual int getB();
	virtual WORD getC();
private:
	int m_B;
	BYTE m_C;
	BYTE m_D;
	WORD m_E;
	DWORD m_F;
	char m_text[16];
}

In the dump window, the user will be able to view the selected address as a class/structure by right-clicking it and choosing "Show as class/strucure", as follows:

; ASCII view
0019AA00 44 33 22 11 BB BB BB BB 66 55 EE  EE FF  FF FF  FF	................
0019AA10 31 32 33 34 35 36 37 38 00 00 00 00 00 00 00 00	12345678........

; Class/Structure view
Address | Type | Field | Value

0019AA00 | void* | vtable | 11223344
0019AA04 | int	 | m_B | BBBBBBBB
0019AA08 | BYTE | m_C | 66
0019AA09 | BYTE | m_D | 55
0019AA0A | WORD | m_E | EEEE
0019AA0C | DWORD | m_F | FFFFFFFF
0019AA10 | char[] | m_text | "12345678" (Ansi)

All the registered structures/classes recognized by the dbg will be available in a separate tab window, so the user can edit them.

**Comment by mrexodia** [comment]
Feel free to implement it, I don't have time right now.

**Comment by lynnux** [comment]
I'v wrote a plugin for ollydbg to do the same thing using liblcang (not published).
I also plan on this for x64dbg, however the pluginsdk lack some feature to achieve this.
Your post inspired me, but I also don't have time right now, I will work on it when I am free.

**Comment by mrexodia** [comment]
Which features are you missing exactly?

**Comment by lynnux** [comment]
The plugin  work like this: 
Popup a window to show the class/struct information. Information in window will be updated by every mouse'click in cpu, memory or stack window. Thus, the plugin should know the focused window(cpu, memory or stack) when mouse clicked. The pluginsdk lack the API to get the focused window.

**Comment by mrexodia** [comment]
@lynnux for now you can easily work around that by adding a menu option in the dump/disasm/stack called 'Struct' or whatever. Another thing you can do is allow users to synchronize the window with an expression (use `DbgValFromString`). You can use `dis.sel()`, `dump.sel()` and `stack.sel()` as expression to get the current selection.

**Comment by blaquee** [comment]
Yea this is well suited for a plugin. @lynnux if there is anything you would like to have implemented for the pluginsdk let it be heard, I have had recent interest in making it better. I have a similar idea for a plugin that resembles an Olly plugin where you can "decode address as" a Windows Structure, so if you know for example a pointer is to be treated as a PEB then you can view it as such...etc..

**Comment by lynnux** [comment]
Thanks guys! Really appreciate yours ideas

**Comment by mrexodia** [comment]
@lynnux I also implemented a plugin to receive click events. From there you can query the selection and see if it changed. See https://github.com/mrexodia/ClickCatcher

**Comment by mrexodia** [comment]
Probably also interesting for @RaMMicHaeL

**Comment by RaMMicHaeL** [comment]
> adding a menu option in the dump/disasm/stack

Is it already possible with the SDK?

**Comment by mrexodia** [comment]
@RaMMicHaeL yeah, but the event filter works again so you can see hotkeys again. I saw you removed that option from multimate assembler.

**Comment by RaMMicHaeL** [comment]
I've mixed it up. For some reason, I though that hotkeys were implemented, and menus weren't, but it's the other way around.
So, it looks like I can implement the hotkeys with the filter API. Should I do that, instead of waiting for a cleaner hotkeys API?

**Comment by mrexodia** [comment]
You can do the filter api or use the favorite commands for now:
http://help.x64dbg.com/en/latest/commands/gui/AddFavouriteCommand.html

I will eventually implement a hotkey api but it has been on my list for
more than two years now so I'm not sure when that will happen 😀

**Comment by mrexodia** [comment]
There has been some progress with this:

![struct](https://i.imgur.com/OxRn83O.png)

**Comment by RaMMicHaeL** [comment]
Handling `WM_KEYUP` in `CBWINEVENT` works. On the other hand, it seems like `CBWINEVENTGLOBAL` stopped working. You can verify it with the existing multiasm version - hotkeys, such as `Ctrl+T` to create a new tab, don't work.

**Comment by mrexodia** [comment]
Yes since qt 5 the global window event does nothing, hotkeys are also in
the local window events now. See the documentation.

**Comment by RaMMicHaeL** [comment]
That worked, thanks. What about the following?
* Keystone and asmjit don't support short jumps, which should be limited to 2 bytes. Works with XEDParse.
* Is there an API for getting the xref analysis information? It can be used for the _Disasm external jumps/calls_ option.

I'm looking at fixing as much as possible and roll out a release of a new version of the plugin.

**Comment by mrexodia** [comment]
@RaMMicHaeL as a preprocessing step I remove `short ` from the string to assemble for keystone. asmjit supposedly supports the `short` prefix but just strips it.

There is an API for xrefs but it only allows queries per address right now, it's called `DbgXrefGet` the operations on that should be O(n) where n is the amount of xrefs on the requested address.

**Comment by RaMMicHaeL** [comment]
I have released a new version of multiasm ([download](http://rammichael.com/downloads/multiasm.rar), [changelog](http://rammichael.com/downloads/multiasm.rar?changelog)), mostly x64db-centric.

Keyboard hotkeys were added, as well as xref support.

> A solution could be to fail if short was included in the instruction and the output was not 2 bytes.

That should work. As x64dbg already strips the string, it would probably be a very simple addition :)

**Comment by mrexodia** [comment]
Most of your requests have been implemented @faltur. I will not add a C++ parser to x64dbg but I will later expose the type system to plugins so they can do it. Additionally there will be a type database (user in the program database and a system database for commonly used types).

Progress so far:

https://gist.github.com/mrexodia/e949ab26d5986a5fc1fa4944ac68147a

![screenshot](https://i.imgur.com/8Rqk0C2.png)

![screenshot2](https://i.imgur.com/dkMDoE7.png)

**Comment by a1ext** [comment]
Is it possible to extract and fill the db of structs/funcs using PDB, like does windbg?

**Comment by mrexodia** [comment]
@a1ext could you elaborate on what you mean? Right now there is no API for
manipulating or enumerating structs if that's what you are talking about.

**Comment by a1ext** [comment]
@mrexodia I mean we may prepare the DB of known structs and functions, which is based on information from PDBs. Then this DB can be used to visualize struct or to show function arguments on the stack.
> Right now there is no API for manipulating or enumerating structs if that's what you are talking about.

Do you mean the x64dbg API?

**Comment by mrexodia** [comment]
> I mean we may prepare the DB of known structs and functions, which is based on information from PDBs. Then this DB can be used to visualize struct or to show function arguments on the stack.

Yeah this is the next step, I will implement a JSON format so we can build a library of structures and functions these will be loaded in x64dbg on startup. The visualization (mainly for functions) is in very basic stages. There is no argument recognition analysis so there isn't much to display.

For plugins we need a good API to enumerate and manipulate the structures but I'm still designing this.
